packages/core/src/api/blockManipulation/blockManipulation.ts

```diff
@@ -141,6 +141,68 @@ export function replaceBlocks<
   blocksToInsert: PartialBlock<BSchema, I, S>[],
   editor: BlockNoteEditor<BSchema, I, S>
 ) {
-  insertBlocks(blocksToInsert, blocksToRemove[0], "before", editor);
-  removeBlocks(blocksToRemove, editor._tiptapEditor);
+  const ttEditor = editor._tiptapEditor;
+  const tr = ttEditor.state.tr;
+
+  const nodesToInsert: Node[] = [];
+  for (const blockSpec of blocksToInsert) {
+    nodesToInsert.push(
+      blockToNode(blockSpec, ttEditor.schema, editor.styleSchema)
+    );
+  }
+
+  const idsOfBlocksToRemove = new Set<string>(
+    blocksToRemove.map((block) =>
+      typeof block === "string" ? block : block.id
+    )
+  );
+
+  let removedSize = 0;
+
+  ttEditor.state.doc.descendants((node, pos) => {
+    // Skips traversing nodes after all target blocks have been removed.
+    if (idsOfBlocksToRemove.size === 0) {
+      return false;
+    }
+
+    // Keeps traversing nodes if block with target ID has not been found.
+    if (
+      node.type.name !== "blockContainer" ||
+      !idsOfBlocksToRemove.has(node.attrs.id)
+    ) {
+      return true;
+    }
+
+    if (
+      node.attrs.id ===
+      (typeof blocksToRemove[0] === "string"
+        ? blocksToRemove[0]
+        : blocksToRemove[0].id)
+    ) {
+      const oldDocSize = tr.doc.nodeSize;
+      tr.insert(pos, nodesToInsert);
+      const newDocSize = tr.doc.nodeSize;
+      removedSize += oldDocSize - newDocSize;
+    }
+
+    idsOfBlocksToRemove.delete(node.attrs.id);
+
+    const oldDocSize = tr.doc.nodeSize;
+    tr.delete(pos - removedSize, pos - removedSize + node.nodeSize);
```

Comment:
is there a reason that for `removeBlocks` we use `BNDeleteBlock` instead of this?

Reply:
Just so we can add the deletion and insertion operations into a single transaction. We don't have a `BNInsertBlocks` function that we can use for the insertion (would need to use `BNCreateBlock` and `BNUpdateBlock`), so it's cleaner to just use PM transactions for both instead of TipTap commands.